### PR TITLE
Add mypy

### LINKED ISSUE #114: Adding mypy checks
### Problem

As the python codebase grows, it might be a good idea to enforce the mypy checks for the consistency of types.

### Solution

- Add mypy dependency
- Add mypy check to Makefile
- Run the check in CI
- Fix all the typing errors (found 180 currently)

Better add it after #109 is complete to prevent duplicated work since it introduced many type changes.

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Fixed #114

### CODE REVIEW

**Comment by icemelon** [comment]
I also suggest changing the [`line-length`](https://github.com/appliedzkp/zkevm-specs/blob/master/pyproject.toml#L9) to 100. I think 120 is a bit too long.

**Comment by ChihChengLiang** [comment]
@icemelon I created #117 for the line-length change. Feel free to give some reviews or comments :)

**Review by ChihChengLiang — COMMENTED** [review]
Hi @han0110  @icemelon,
I have some questions about the instruction part. Not sure what do you think about them.

File: src/zkevm_specs/evm/instruction.py [hunk 1]
```diff
@@ -270,10 +276,6 @@ def mul_word_by_u64(self, multiplicand: RLC, multiplier: FQ) -> Tuple[RLC, int]:
     def rlc_to_le_bytes(self, rlc: RLC) -> bytes:
         return rlc.le_bytes
 
-    def rlc_to_int_unchecked(self, rlc: RLC, n_bytes: int) -> int:
```
[the remark this answers is not in the document]
Comment (icemelon) [reply src/zkevm_specs/evm/instruction.py:281]: I think we still need to have this. It's used when computing the witness data and can allow overflow

File: src/zkevm_specs/evm/instruction.py [hunk 2]
```diff
@@ -616,22 +614,22 @@ def transfer(
         )
         return sender_balance_pair, receiver_balance_pair
 
-    def memory_offset_and_length_to_int(self, offset: RLC, length: RLC) -> Tuple[int, int]:
-        length = self.rlc_to_int_exact(length, N_BYTES_MEMORY_ADDRESS)
-        if self.is_zero(length):
+    def memory_offset_and_length_to_int(self, _offset: RLC, _length: RLC) -> Tuple[int, int]:
```
[the remark this answers is not in the document]
Comment (icemelon) [reply src/zkevm_specs/evm/instruction.py:635]: Yes, I have a few more fixes in the #111

**Comment by ChihChengLiang** [comment]
This task is blocked. We are constantly juggling 3 types:

- IntEnum
- FQ
- RLC

Usually, a cell in a table would be of different types. For example, a value of TxTable could be gasPrice, which is stored in RLC. A value could also be `caller_address`, which is an FQ.

This makes our txTable lookup outputs Union[RLC, FQ], and we later need to eliminate the type to 1 possibility, which is not Python really good at.

We keep RLC because usually, we want RLC.le_bytes for computation. But the RLC.value is what is acutally stored.

File: src/zkevm_specs/evm/instruction.py [hunk 3]
```diff
@@ -314,124 +316,120 @@ def bytes_range_lookup(self, value: FQ, n_bytes: int) -> Sequence[int]:
...
 
-        return self.tables.rw_lookup([rw_counter, rw, tag] + inputs)
-
-    def state_write_only_persistent(
-        self,
-        tag: RWTableTag,
-        inputs: Sequence[int],
-        is_persistent: bool,
-    ) -> Array10:
-        assert tag.write_only_persistent()
```
[the remark this answers is not in the document]
Comment (han0110) [reply src/zkevm_specs/evm/instruction.py:378]: No, I think we forgot to remove this in previous PR.

**Comment by ChihChengLiang** [comment]
close in favor of @han0110's new approach. PR would be submitted soon after recent spec PRs are merged.

**Comment by ed255** [comment]
> This task is blocked. We are constantly juggling 3 types:
> 
>     * IntEnum
> 
>     * FQ
> 
>     * RLC
> 
> 
> Usually, a cell in a table would be of different types. For example, a value of TxTable could be gasPrice, which is stored in RLC. A value could also be `caller_address`, which is an FQ.
> 
> This makes our txTable lookup outputs Union[RLC, FQ], and we later need to eliminate the type to 1 possibility, which is not Python really good at.
> 
> We keep RLC because usually, we want RLC.le_bytes for computation. But the RLC.value is what is acutally stored.

This sounds like a case where duck typing is needed, and for the typing we should use something like an Interface.  I've taken a look at python support for "Interfaces" and there's `Protocol` which allows expressing types that must implement some methods.
See https://stackoverflow.com/a/60766965 for an example, and see for the docs https://stackoverflow.com/a/60766965

So we could have something like
```
class Expression(Protocol):
    def value(self) -> FQ:
        ...
```
And have `IntEnum`, `FQ` and `RLC` implement the method `value`

**Pull request closed without merging**